src/lib/storage/table.cpp

```diff
@@ -203,6 +213,68 @@ TableType Table::get_type() const {
   }
 }
 
+void Table::set_partitioning_and_clear(std::shared_ptr<AbstractPartitionSchema> partition_schema) {
+  _partition_schema = partition_schema;
+  _chunks.clear();
+}
+
+void Table::apply_partitioning(std::shared_ptr<AbstractPartitionSchema> partition_schema) {
+  _partition_schema = partition_schema;
+  create_initial_chunks(static_cast<PartitionID>(partition_schema->partition_count()));
+}
+
+void Table::create_hash_partitioning(const ColumnID column_id, const HashFunction hash_function,
+                                     const size_t number_of_partitions) {
+  _partition_schema = std::make_shared<HashPartitionSchema>(column_id, hash_function, number_of_partitions);
+  create_initial_chunks(static_cast<PartitionID>(number_of_partitions));
+}
+
+void Table::create_null_partitioning() {
+  _partition_schema = std::make_shared<NullPartitionSchema>();
+  create_new_chunk(PartitionID{0});
+}
+
+void Table::create_range_partitioning(const ColumnID column_id, const std::vector<AllTypeVariant> bounds) {
+  _partition_schema = std::make_shared<RangePartitionSchema>(column_id, bounds);
+  create_initial_chunks(static_cast<PartitionID>(bounds.size() + 1));
+}
+
+void Table::create_round_robin_partitioning(const size_t number_of_partitions) {
+  _partition_schema = std::make_shared<RoundRobinPartitionSchema>(number_of_partitions);
+  create_initial_chunks(static_cast<PartitionID>(number_of_partitions));
+}
+
+void Table::create_initial_chunks(PartitionID number_of_partitions) {
+  if (row_count() > 0) {
+    // the initial chunk was not used yet
+    throw std::runtime_error("Unable to create partitioning on non-empty table");
+  }
+  _chunks.clear();
+  for (auto partition_id = PartitionID{0}; partition_id < number_of_partitions; ++partition_id) {
+    create_new_chunk(partition_id);
+  }
+}
+
+bool Table::is_partitioned() const { return _partition_schema->is_partitioned(); }
+
+uint16_t Table::partition_count() const { return _partition_schema->partition_count(); }
+
+ChunkID Table::get_chunk_id(const std::shared_ptr<Chunk> chunk) const {
+  ptrdiff_t pos = std::distance(
+      _chunks.begin(), std::find_if(_chunks.begin(), _chunks.end(),
```

Comment:
this seems awfully expensive on large tables. could we store the ids somehow? if not, at least comment this because of O(n) lookup